### PR TITLE
Feat/856 Validate Reference Radiation IDs

### PR DESCRIPTION
**Pull request by demariadaniel**
**Description of changes**

https://app.zenhub.com/workspaces/icgc-argo-platform-dk-production-board-5e542d38415f5034e9fed89d/issues/gh/icgc-argo/argo-clinical/856

**Type of Change**

- [ ] Bug
- [ ] Refactor
- [x] New Feature
- [ ] Release Candidate

**Checklist before requesting review:**

- [ ] Check branch (code change PRs go to `develop` not master)
- [ ] Check copyrights for new files
- [ ] Manual testing
- [ ] Regression tests completed and passing (double check number of tests).
- [ ] Spelling has been checked.
- [ ] Updated swagger docs accordingly (check it's still valid)
- [ ] Set `validationDependency` in meta tag for [Argo Dictionary](https://github.com/icgc-argo/argo-dictionary) fields used in code

### CODE REVIEW

File: src/submission/validation-clinical/therapy.ts
Comment (ciaranschutte) [inline src/submission/validation-clinical/therapy.ts:61]: `treatment_type` will always have `includes` property because its core js array api and the previous line makes sure its an array

File: src/submission/validation-clinical/therapy.ts
Comment (ciaranschutte) [inline src/submission/validation-clinical/therapy.ts:72]: this function name could be more descriptive. similar to the other funcs eg. "checkThisThingForThisThing"

File: src/submission/submission-error-messages.ts
Comment (ciaranschutte) [inline src/submission/submission-error-messages.ts:121]: might be worth sending back the submitted treatment ids too. i believe its available in the data argument

File: src/submission/validation-clinical/therapy.ts
Comment (ciaranschutte) [reply src/submission/validation-clinical/therapy.ts:61]: i misread this. this is checking if `treatment_type` is defined? which previous line should cover

**Comment by ciaranschutte** [comment]
Ah I see this is a WIP. Confused it with the other one that's waiting on a review. Maybe we could do "draft pull requests" for WIP that's up as a pr

File: src/submission/validation-clinical/therapy.ts
Comment (demariadaniel) [reply src/submission/validation-clinical/therapy.ts]: Yes I've updated to `validateRadiationRecords`

File: src/submission/validation-clinical/therapy.ts
Comment (demariadaniel) [reply src/submission/validation-clinical/therapy.ts]: True, updated

File: src/submission/validation-clinical/therapy.ts
Comment (demariadaniel) [reply src/submission/validation-clinical/therapy.ts]: So this, I was following the pattern here:
https://github.com/icgc-argo/argo-clinical/blob/master/src/submission/validation-clinical/index.ts#L36

And thinking we should maybe export `radiation` as it's own collection of validators the way we do with other entities. But maybe I'll just leave it as part of `treatment` for now rather than mess with the existing setup.

File: src/submission/submission-service.ts
Comment (demariadaniel) [inline src/submission/submission-service.ts:359]: Updated spelling errors

File: src/submission/validation-clinical/therapy.ts
Comment (demariadaniel) [inline src/submission/validation-clinical/therapy.ts:155]: This final validation needs one last pass to get working

File: src/submission/submission-error-messages.ts
Comment (demariadaniel) [reply src/submission/submission-error-messages.ts:119]: This is now added!

File: src/submission/validation-clinical/surgery.ts
Comment (demariadaniel) [inline src/submission/validation-clinical/surgery.ts:38]: More spelling

File: src/submission/validation-clinical/therapy.ts
Comment (demariadaniel) [reply src/submission/validation-clinical/therapy.ts:226]: Updated, tested + working

File: src/submission/validation-clinical/therapy.ts
Comment (demariadaniel) [reply src/submission/validation-clinical/therapy.ts]: OK so this has changed quite a bit
First, radiation function is now `validateRadiationRecords`
Second, we were previously only validating a single record for a single donor at a time, so more important than separating radiation functions was to create a place for cross file validators
See `src/submission/validation-clinical/therapy.ts @ 70`
This should likely be moved out of therapy into another file but started here

**Comment by ciaranschutte** [comment]
is there by any chance a list of files that have changes other spelling fixes? just wondering before I review comb through as I don't want to miss anything thinking its a speller @demariadaniel

File: src/submission/validation-clinical/therapy.ts
Comment (ciaranschutte) [inline src/submission/validation-clinical/therapy.ts:113]: is this renaming to provide more context in the file? ie. in the whole system it's `submitter_donor_id` but we are using it here as `submittedTherapyDonorId`?

File: src/submission/validation-clinical/therapy.ts
Comment (ciaranschutte) [inline src/submission/validation-clinical/therapy.ts:125]:
``` typescript

  for (const donorId in submittedRecords) {
    const submittedTreatmentRecords = submittedRecords[donorId].treatment
    if (submittedTreatmentRecords?.length)
      submittedTreatments = [...submittedTreatments, ...submittedTreatmentRecords];
  }
  ```
  
  I think this works the same (ignore typings for now).
  `submittedRecord[donorId].treatment` looks like it should be an array already, so the spread new array thing is redundant

File: src/submission/validation-clinical/therapy.ts
Comment (ciaranschutte) [inline src/submission/validation-clinical/therapy.ts:138]: I think this is not working as expected. likely the `donorDao` is always returning an array of some kind - TS will tell you the type of `{ donors }` is `const donors: readonly DeepReadonly<Donor>[]`

so `emptyDonorResult` is never assigned and not needed

File: src/submission/validation-clinical/therapy.ts
Comment (ciaranschutte) [inline src/submission/validation-clinical/therapy.ts:145]: you can save an iteration of the array by just using a `filter`
most array methods will give you back empty arrays if operating on empty array 
ie. they won't throw errors for empty arrays
`[].filter(x => x?.y?.z).flat() // result is []`


I think this could be simplified to:
```typescript
  const storedTreatments = donors.filter(donor => Boolean(donor?.treatments)).flat();
```

File: src/submission/validation-clinical/therapy.ts
Comment (ciaranschutte) [inline src/submission/validation-clinical/therapy.ts:183]: ` const currentTreatmentIsRadiation =  treatment_type === 'radiation';`
not doing any property lookup , so no need to check if the object exists first

File: src/submission/validation-clinical/therapy.ts
Comment (ciaranschutte) [inline src/submission/validation-clinical/therapy.ts:189]: may as well just stick with optional chaining for the full block rather than half old school object check

**Review by ciaranschutte — CHANGES_REQUESTED** [review]
awesome work ! - thank you for all the clean up too! good separation of logic and handling. couple of syntaxy js suggestions. will take a longer look tomorrow to get my head around the business logic. @joneubank would likely be a better eye for that - suggestining 2nd reviewer before merge

File: src/submission/validation-clinical/therapy.ts
Comment (demariadaniel) [reply src/submission/validation-clinical/therapy.ts:56]: Great points!

File: src/submission/validation-clinical/therapy.ts
Comment (demariadaniel) [reply src/submission/validation-clinical/therapy.ts:85]: The idea here is that it is only equal to `radiationErrors` until we add more validations
I try not to spend much time solving 'future problems' that might not exist, but my next ~3 tickets are all cross-file validation
So this will look more like:
`const crossFileErrors = [...radiationErrors, ...followUpErrors, ...chemotherapyErrors];`
Given that perspective, let me know if there's a better approach

File: src/submission/validation-clinical/therapy.ts
Comment (demariadaniel) [reply src/submission/validation-clinical/therapy.ts:125]: I will give a second look, given I used 'as' this is definitely a crutch specifically to manage Types

File: src/submission/validation-clinical/therapy.ts
Comment (demariadaniel) [reply src/submission/validation-clinical/therapy.ts:138]: Worth a review, I added this in response to unit test timeout errors, now that I figured out the real solution this may be unneeded

File: src/submission/validation-clinical/therapy.ts
Comment (demariadaniel) [reply src/submission/validation-clinical/therapy.ts:113]: Yes. The intent here is to disambiguate between the Submitter Donor ID associated with the Therapy Record we are currently validating, and any of the multiple Submitter Donor IDs we are matching it against

We are checking, does Submitter Donor ID A match Submitter Donor ID B, C, D, etc., so using the key name is not important necessarily, the end result is just a boolean

Comment (demariadaniel) [reply src/submission/validation-clinical/therapy.ts:113]: There might be potential to just lump all comparison values together, i.e. we have our `submitter_donor_id` which is from the submitted therapy record, then we loop through a new `relatedTreatments[]` and compare if `submitter_donor_id === relatedTreatments[i].submitter_donor_id`, does that make sense?

File: src/submission/validation-clinical/therapy.ts
Comment (demariadaniel) [reply src/submission/validation-clinical/therapy.ts:183]: I see, yes, will update

File: src/submission/validation-clinical/therapy.ts
Comment (ciaranschutte) [reply src/submission/validation-clinical/therapy.ts:85]: Great! Yes this just didn't click for me on reading - makes sense as is!

File: src/submission/validation-clinical/therapy.ts
Comment (ciaranschutte) [reply src/submission/validation-clinical/therapy.ts:125]: i don't think the `as` assertion is needed , submittedRecords is typed already so submittedTreatmentRecords should be grand as is

File: src/submission/validation-clinical/therapy.ts
Comment (demariadaniel) [reply src/submission/validation-clinical/therapy.ts]: Updated

File: src/submission/validation-clinical/therapy.ts
Comment (demariadaniel) [reply src/submission/validation-clinical/therapy.ts:138]: Aha I made this work using `sinon.stub.resolves` in the unit test, and reverted this line to `const { donors } = await donorDao.findByPaginatedProgramId(programId, query);`

File: src/submission/validation-clinical/therapy.ts
Comment (demariadaniel) [reply src/submission/validation-clinical/therapy.ts:80]: Updated

File: src/submission/validation-clinical/therapy.ts
Comment (demariadaniel) [reply src/submission/validation-clinical/therapy.ts]: Removed

File: src/submission/validation-clinical/therapy.ts
Comment (demariadaniel) [reply src/submission/validation-clinical/therapy.ts:108]: Keeping as is for now, can revisit a map solution at a later time

File: src/submission/validation-clinical/therapy.ts
Comment (demariadaniel) [reply src/submission/validation-clinical/therapy.ts]: Updated

File: src/submission/validation-clinical/therapy.ts
Comment (demariadaniel) [reply src/submission/validation-clinical/therapy.ts:120]: Yep looks like this works

**Review by ciaranschutte — APPROVED** [review]

**Pull request merged**
